### PR TITLE
Issue 46427: Slow queries for runs related to datas and materials

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
We can avoid joining to the `exp.data` and `exp.material` tables for these queries because they're only invoked for single rows at a time, and we already have the data object so we know if it has a `sourceApplicationId`

#### Changes
* Drop the extra join and don't do an `OR` unless we need to

### CODE REVIEW

**Review by labkey-susanh — APPROVED** [review]

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 1]
```diff
@@ -5458,33 +5458,33 @@ public List<ExpRunImpl> getRunsUsingDataIds(List<Integer> ids)
...
                         INNER JOIN exp.ProtocolApplication pa ON pa.RunId = r.RowId
                         INNER JOIN exp.DataInput di ON di.targetApplicationId = pa.RowId
-                        LEFT OUTER JOIN exp.Data d ON pa.RowId = d.sourceApplicationID
-                        WHERE di.dataId\s""");
-        getExpSchema().getSqlDialect().appendInClauseSql(sql, ids);
-        sql.append("\n");
-        sql.append("OR d.RowId ");
-        getExpSchema().getSqlDialect().appendInClauseSql(sql, ids);
-        sql.append("\n");
+                        WHERE di.dataId = ?""");
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]:
```suggestion
                        WHERE di.dataId = ? """);
```

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 2]
```diff
@@ -5458,33 +5458,33 @@ public List<ExpRunImpl> getRunsUsingDataIds(List<Integer> ids)
...
+        sql.add(data.getRowId());
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Should this use `sql.appendValue` instead?

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 3]
```diff
@@ -5531,33 +5531,33 @@ private SQLFragment getAppendInClause(Collection<?> ids)
...
                         INNER JOIN exp.MaterialInput mi ON mi.targetApplicationId = pa.RowId
-                        LEFT OUTER JOIN exp.Material m ON pa.RowId = m.sourceApplicationID
-                        WHERE mi.materialId\s""");
-        getExpSchema().getSqlDialect().appendInClauseSql(sql, ids);
-        sql.append("\n");
-        sql.append("OR m.RowId ");
-        getExpSchema().getSqlDialect().appendInClauseSql(sql, ids);
-        sql.append("\n");
+                        WHERE mi.materialId = ?""");
+        sql.add(material.getRowId());
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Same comments here.

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java (hunk 1 above)
Comment (labkey-jeckels) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: I fixed the whitespace problem with a different commit.

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java (hunk 2 above)
Comment (labkey-jeckels) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: `appendValue()` inlines the value into the SQL itself. `add()` supplies a value for the `?` parameter. The latter means that the DB can reuse the same query plan over multiple executions, and that we don't end up with a bunch of variations with different rowId values in the query profiler.

Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Ah right.

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 4]
```diff
@@ -5531,33 +5531,33 @@ private SQLFragment getAppendInClause(Collection<?> ids)
...
-        sql.append("\n");
-        sql.append("OR m.RowId ");
-        getExpSchema().getSqlDialect().appendInClauseSql(sql, ids);
-        sql.append("\n");
-        sql.append("ORDER BY r.Created DESC");
+                        WHERE mi.materialId = ?""");
+        sql.add(material.getRowId());
+        if (material.getSourceApplication() != null)
+        {
+            sql.append(" OR pa.RowId = ?");
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: OR where the two sides reference different tables are _terrible_
Better to just UNION the two cases

**Review by labkey-matthewb — CHANGES_REQUESTED** [review]

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java (hunk 4 above)
Comment (labkey-jeckels) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: OK. New version pushed.

**Review by labkey-matthewb — APPROVED** [review]

**Pull request merged**